linkis-engineconn-plugins/hbase/src/main/java/org/apache/linkis/manager/engineplugin/hbase/HBaseConnectionManager.java

```diff
@@ -0,0 +1,358 @@
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.HBASE_SECURITY_AUTH;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.KERBEROS;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.KERBEROS_KEYTAB_FILE;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.KERBEROS_PRINCIPAL;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.KERBEROS_PROXY_USER;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.KRB5_CONF_PATH;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.MASTER_SERVER_KERBEROS_PRINCIPAL;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.REGION_SERVER_KERBEROS_PRINCIPAL;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.SIMPLE;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.UNIQUE_KEY_DELIMITER;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.ZOOKEEPER_CLIENT_PORT;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.ZOOKEEPER_NODE_PARENT;
+import static org.apache.linkis.manager.engineplugin.hbase.constant.HBaseEngineConnConstant.ZOOKEEPER_QUORUM;
+
+public class HBaseConnectionManager {
+  private static final Logger LOG = LoggerFactory.getLogger(HBaseConnectionManager.class);
+  private final ConcurrentHashMap<String, Connection> connectionMap;
+  private final ReentrantLock lock = new ReentrantLock();
+  private static final AtomicBoolean kerberosEnvInit = new AtomicBoolean(false);
+  private static final int KERBEROS_RE_LOGIN_MAX_RETRY = 5;
+  private static final long KERBEROS_RE_LOGIN_INTERVAL = 30 * 60 * 1000L;
+  private static volatile HBaseConnectionManager instance = null;
+
+  private HBaseConnectionManager() {
+    connectionMap = new ConcurrentHashMap<>();
+  }
+
+  public static HBaseConnectionManager getInstance() {
+    if (instance == null) {
+      synchronized (HBaseConnectionManager.class) {
+        if (instance == null) {
+          instance = new HBaseConnectionManager();
+        }
+      }
+    }
+    return instance;
+  }
+
+  public Configuration getConfiguration(Properties prop) {
+    Map<String, String> propMap = getPropMap(prop);
+    return buildConfiguration(propMap);
+  }
+
+  public Connection getConnection(Properties prop) {
+    Map<String, String> propMap = getPropMap(prop);
+    return getConnection(propMap);
+  }
+
+  public Connection getConnection(Configuration configuration) {
+    String clusterConnUniqueKey = generateUniqueConnectionKey(configuration);
+    LOG.info("Start to get connection for cluster {}.", clusterConnUniqueKey);
+    try {
+      lock.lock();
```

Comment:
Is it better to first determine whether it contains the corresponding key and then lock it?